compiler/qsc_eval/src/lib.rs

```diff
@@ -424,44 +491,70 @@ impl State {
 
     /// # Errors
     /// Returns the first error encountered during execution.
-    pub fn resume(
+    /// # Panics
+    /// When returning a value in the middle of execution.
+    pub fn eval(
         &mut self,
         globals: &impl NodeLookup,
         env: &mut Env,
         sim: &mut impl Backend<ResultType = impl Into<val::Result>>,
         out: &mut impl Receiver,
         breakpoints: &[StmtId],
-    ) -> Result<Option<StmtId>, (Error, Vec<Frame>)> {
+        step: &StepAction,
+    ) -> Result<StepResult, (Error, Vec<Frame>)> {
+        let current_frame = self.call_stack.len();
+
         while let Some(cont) = self.pop_cont() {
             let res = match cont {
-                Cont::Action(action) => self
-                    .cont_action(env, sim, globals, action, out)
-                    .map(|_| None),
-                Cont::Expr(expr) => self.cont_expr(env, globals, expr).map(|_| None),
+                Cont::Action(action) => {
+                    self.cont_action(env, sim, globals, action, out)
+                        .map_err(|e| (e, self.get_stack_frames()))?;
+                    continue;
+                }
+                Cont::Expr(expr) => {
+                    self.cont_expr(env, globals, expr)
+                        .map_err(|e| (e, self.get_stack_frames()))?;
+                    continue;
+                }
                 Cont::Frame(len) => {
                     self.leave_frame(len);
-                    Ok(None)
+                    continue;
                 }
                 Cont::Scope => {
                     env.leave_scope();
-                    Ok(None)
+                    continue;
                 }
                 Cont::Stmt(stmt) => {
                     self.cont_stmt(globals, stmt);
-                    match breakpoints.iter().find(|&bp| *bp == stmt) {
-                        Some(bp) => Ok(Some(bp)),
-                        None => Ok(None),
+                    if let Some(bp) = breakpoints.iter().find(|&bp| *bp == stmt) {
+                        StepResult::BreakpointHit(*bp)
+                    } else {
+                        // no breakpoint, but we may stop here
+                        if matches!(step, StepAction::In) {
+                            StepResult::StepIn
+                        } else if matches!(step, StepAction::Next)
+                            && current_frame == self.call_stack.len()
+                        {
+                            StepResult::Next
+                        } else if matches!(step, StepAction::Out)
+                            && current_frame > self.call_stack.len()
+                        {
+                            StepResult::StepOut
+                        } else {
+                            continue;
+                        }
```

Comment:
Since `StepResult` is a simple enum, these could use equality comparison instead of matches if you derive `PartialEq`, ie:

```suggestion
                        if step == StepAction::In {
                            StepResult::StepIn
                        } else if step == StepAction::Next
                            && current_frame == self.call_stack.len()
                        {
                            StepResult::Next
                        } else if step == StepAction::Out
                            && current_frame > self.call_stack.len()
                        {
                            StepResult::StepOut
                        } else {
                            continue;
                        }
```